File: fsharp/grep/Grep.fs
module Grep

open System
open System.Text.RegularExpressions
open System.IO

[<Flags>]
type GrepFlags = 
    | None =            0uy 
    | LineNums =        1uy 
    | OnlyFileNames =   2uy 
    | CaseInsensitive = 4uy
    | InvertMatch =     8uy
    | FullLineMatch =   16uy

let parseFlag (flag: string) = 
    match flag with 
    | "-n" -> GrepFlags.LineNums 
    | "-l" -> GrepFlags.OnlyFileNames 
    | "-i" -> GrepFlags.CaseInsensitive
    | "-v" -> GrepFlags.InvertMatch
    | "-x" -> GrepFlags.FullLineMatch
    |  _   -> GrepFlags.None 

let parseFlags (flagArguments: string list) = 
    flagArguments |> List.fold (fun acc flag -> acc ||| (parseFlag flag)) GrepFlags.None 
    
let format (use_files: bool) (use_nums: bool) (file: string) (n: int) (line: string) = 
    let file_format = if use_files then $"{file}:" else ""
    let num_format = if use_nums then $"{n}:" else ""
    $"{file_format}{num_format}{line}"

// no logical exclusive or operator for booleans in F# !!
let xor (a: bool) (b: bool) = 
    (a && not b) || (not a && b)

let grep (files: string list) (flagArguments: string list) (pattern: string) = 
    let flags = parseFlags flagArguments 
    let pattern_str = if flags.HasFlag GrepFlags.FullLineMatch then $"^{pattern}$" else pattern 
    let rgx_options = if flags.HasFlag GrepFlags.CaseInsensitive then RegexOptions.IgnoreCase else RegexOptions.None
    let rgx = new Regex(pattern_str, rgx_options)
    if flags.HasFlag GrepFlags.OnlyFileNames 
        then 
            files 
            |> List.filter (fun file -> 
                 File.ReadAllLines(file) 
                 |> Seq.exists(fun line -> 
                    xor (rgx.IsMatch(line)) (flags.HasFlag GrepFlags.InvertMatch)
                 )
            )
        else 
            files 
            |> List.collect (fun file -> 
                 File.ReadAllLines(file)
                 |> Seq.indexed
                 |> Seq.filter (fun (_, line) -> 
                    xor (rgx.IsMatch(line)) (flags.HasFlag GrepFlags.InvertMatch )
                 )
                 |> Seq.map (fun (n, line) -> 
                    format (files.Length > 1) (flags.HasFlag GrepFlags.LineNums) file (n+1) line
                 )
                 |> Seq.toList
            )

File: csharp/grep/Grep.cs
using System;
using System.IO;
using System.Text.RegularExpressions;

public static class Grep
{
    static string Format(bool use_filename, bool use_line_number, string filename, int line_number, string val)
    {
        string file = null;
        if (use_filename)
        {
            file = filename + ':';
        } else
        {
            file = string.Empty;
        }
        string num = null;
        if (use_line_number)
        {
            num = line_number.ToString() + ':';
        } else
        {
            num = string.Empty;
        }
        return $"{file}{num}{val}\n";
    }
    public static string Match(string pattern, string flags, string[] files)
    {
        string result = string.Empty;
        string pattern_str = null;
        if (flags.Contains("-x"))
        {
            pattern_str = $"^{pattern}$";
        } else
        {
            pattern_str = pattern;
        }
        Regex rgx = null;
        if (flags.Contains("-i"))
        {
            rgx = new Regex(pattern_str, RegexOptions.IgnoreCase | RegexOptions.Compiled);
        } else
        {
            rgx = new Regex(pattern_str, RegexOptions.Compiled);
        }
        foreach (string file in files)
        {
            using (StreamReader sreader = File.OpenText(file))
            {
                string line = null;
                int line_number = 0;
                while ((line = sreader.ReadLine())!= null)
                {
                    line_number++;
                    if (rgx.IsMatch(line) ^ flags.Contains("-v"))
                    {
                        if (flags.Contains("-l"))
                        {
                            result += file + '\n';
                            break;
                        } else
                        {
                            result += Format(files.Length > 1, flags.Contains("-n"), file, line_number, line);
                        }
                    }
                }
            }
        }
        return result.TrimEnd();
    }
}